### PR TITLE
[otbn] Add ECC to register files and call stack

### LINKED ISSUE #6186: [otbn] Implement OTBN.H.REG_INTEGRITY
Write RTL for it.

> estimate 16

### PR DESCRIPTION
**Pull request by GregAC**
ECC SECDED is applied to both base and bignum register files. For the
base register file a single ECC is applied to whole 32-bit word. For the
bignum register file there is a seperate ECC per 32-bits (8 in all to
cover a 256-bit register).

ECC encode and decode is implemented by the register file. No error
correction is applied, ECCs are only used for error detection.

The register file interfaces are expanded to allow a direct ECC write.
This is used where whatever is supplying the data already has an ECC so
it doesn't need to be recalculated (e.g. during a bignum register move
or select).

Fixes #6186

Signed-off-by: Greg Chadwick <gac@lowrisc.org>

### CODE REVIEW

File: hw/ip/otbn/rtl/otbn_rf_base_ff.sv [hunk 1]
```diff
@@ -17,16 +20,16 @@ module otbn_rf_base_ff
 
   input logic [4:0]    wr_addr_i,
   input logic          wr_en_i,
-  input logic [31:0]   wr_data_i,
+  input logic [38:0]   wr_data_i,
```
Comment (tjaychen) [inline hw/ip/otbn/rtl/otbn_rf_base_ff.sv:23]: does it make sense to parameterize this `width` just in case we change algorithms in the future?

File: hw/ip/otbn/rtl/otbn_rf_bignum.sv [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      .rd_addr_a_i,
+      .rd_data_a_o(rd_data_ecc_a_o),
+
+      .rd_addr_b_i,
+      .rd_data_b_o(rd_data_ecc_b_o)
+    );
+  end
+
+  // New data can have its ECC from an external source or the ECC can be calculated here
+  assign wr_data_ecc = wr_data_ecc_direct_sel_i ? wr_data_ecc_direct_i : wr_data_ecc_calc;
```
Comment (tjaychen) [inline hw/ip/otbn/rtl/otbn_rf_bignum.sv:87]: just a note for a separate PR. 
We probably want to consider hardening the `sel` signal.

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 3]
```diff
@@ -297,7 +311,7 @@ module otbn_controller
   end
 
   // No RF integrity checks implemented yet
-  assign err_bits_o.fatal_reg     = 1'b0;
+  assign err_bits_o.fatal_reg     = rf_base_rd_data_err_i | rf_bignum_rd_data_err_i;
```
Comment (tjaychen) [inline hw/ip/otbn/rtl/otbn_controller.sv:314]: is the above comment out of date?

Comment (tjaychen) [reply hw/ip/otbn/rtl/otbn_controller.sv:314]: also as a side note for the future, we probably want to consider hardening the error signal here as well.

**Comment by tjaychen** [comment]
thanks @GregAC ! let me know if you want to try and get this in for silver.
If not that's fine too, since we do have area estimates for the eventual `otbn` anyhow.

File: hw/ip/otbn/rtl/otbn_rf_bignum.sv [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * applied.
+ *
+ * `wr_data_i` supplies data that requires ECC calulation and `wr_data_ecc_direct_i` supplies data
+ * that comes with ECC. `wr_data_ecc_direct_sel_i` is asserted to select the data with ECC for the
+ * write, otherwise ECC is calculated seperately from `wr_data_i`.
+ *
+ * Features:
+ * - 2 read ports
+ * - 1 write port
+ * - ECC protection
```
Comment (cdgori) [inline hw/ip/otbn/rtl/otbn_rf_bignum.sv:22]: Maybe rephrase this to "ECC for double-error-detection (but no correction)"

File: hw/ip/otbn/rtl/otbn_rf_bignum_ff.sv [hunk 5]
```diff
@@ -3,7 +3,10 @@
 // SPDX-License-Identifier: Apache-2.0
 
 /**
- * WLEN (256b) Wide Register File (WDRs)
+ * ExtWLEN (312b) Wide Register File (WDRs)
+ *
+ * ExtWLEN allows bits to provide ECC protection to WLEN words on a 32-bit granule. ECC
```
Comment (cdgori) [inline hw/ip/otbn/rtl/otbn_rf_bignum_ff.sv:8]: Change "ECC protection" to "integrity checking via ECC" (?)

File: hw/ip/otbn/rtl/otbn_rf_bignum_fpga.sv [hunk 6]
```diff
@@ -3,7 +3,10 @@
 // SPDX-License-Identifier: Apache-2.0
 
 /**
- * WLEN (256b) Wide Register File (WDRs)
+ * ExtWLEN (312b) Wide Register File (WDRs)
+ *
+ * ExtWLEN allows bits to provide ECC protection to WLEN words on a 32-bit granule. ECC
```
Comment (cdgori) [inline hw/ip/otbn/rtl/otbn_rf_bignum_fpga.sv:8]: Change "ECC protection" to "integrity checking via ECC" (?)

**Review by cdgori — APPROVED** [review]
Just some wording nits for comments, LGTM.

File: hw/ip/otbn/rtl/otbn_rf_bignum.sv (hunk 2 above)
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_rf_bignum.sv:87]: We certainly need to look at some control signal hardening in general. I wonder if we can do something such that when `wr_data_ecc_direct_sel_i` is set `wr_data_ecc_calc` is some invalid value (i.e. would fail an ECC check)?

File: hw/ip/otbn/rtl/otbn_controller.sv (hunk 3 above)
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_controller.sv:314]: Ah yes, well spotted

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 7]
```diff
@@ -59,17 +59,24 @@ module otbn_controller
...
+  output logic               rf_bignum_rd_en_a_o,
+  input  logic [ExtWLEN-1:0] rf_bignum_rd_data_ecc_a_i,
 
-  output logic [4:0]      rf_bignum_rd_addr_b_o,
-  input  logic [WLEN-1:0] rf_bignum_rd_data_b_i,
+  output logic [4:0]         rf_bignum_rd_addr_b_o,
+  output logic               rf_bignum_rd_en_b_o,
+  input  logic [ExtWLEN-1:0] rf_bignum_rd_data_ecc_b_i,
+
+  input  logic            rf_bignum_rd_data_err_i,
```
Comment (imphil) [inline hw/ip/otbn/rtl/otbn_controller.sv:79]: nit: indentation

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 8]
```diff
@@ -543,13 +574,26 @@ module otbn_controller
   assign mac_bignum_rf_wr_data[WLEN/2-1:0] = mac_bignum_operation_result_i[WLEN/2-1:0];
 
   always_comb begin
+    // Write data mux for anything that needs ECC computing during register write
     unique case (insn_dec_bignum_i.rf_wdata_sel)
       RfWdSelEx:   rf_bignum_wr_data_o = alu_bignum_operation_result_i;
       RfWdSelLsu:  rf_bignum_wr_data_o = lsu_bignum_rdata_i;
       RfWdSelIspr: rf_bignum_wr_data_o = ispr_rdata_i;
```
Comment (imphil) [inline hw/ip/otbn/rtl/otbn_controller.sv:581]: These two shouldn't need ECC computation going forward. Can you add a TODO?

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 9]
```diff
@@ -543,13 +574,26 @@ module otbn_controller
...
+
+    // Write data mux for anything that provides its own ECC
+    unique case (insn_dec_bignum_i.rf_wdata_sel)
+      RfWdSelMovSel: begin
+        rf_bignum_wr_data_ecc_direct_sel_o = 1'b1;
+        rf_bignum_wr_data_ecc_direct_o     = selection_result;
+      end
+      default: begin
+        rf_bignum_wr_data_ecc_direct_sel_o = 1'b0;
+        rf_bignum_wr_data_ecc_direct_o     = selection_result;
```
Comment (imphil) [inline hw/ip/otbn/rtl/otbn_controller.sv:594]: Should we make this 0 by default right from the start? (We plan to do the datapath blanking anyways.)

**Review by imphil — APPROVED** [review]
Thanks Greg! Looks good with some small comments. I extensively stared at the block diagram and I think we got the relevant areas covered (thanks for also thinking about the REG-REG moves and selects!); the full picture will only unfold once we add further integrity protection.

* It would be great if we could use signal names that indicate clearly if they are protected or not. E.g. `rf_data_protected` vs `rf_data_unprotected` or something like that (probably needs to be shorter, `intg` is the short version of integrity that we have been using elsewhere).
* The comments repeatedly mention the integrity protection algorithm and its properties. I'd just remove them for now and point to the documentation, I have a PR upcoming there shortly. In particular, we guarantee the detection of three error bits per 32 bit word (not 2).
* Naming: I find "ECC" always confusing in the context here. Yes, we use an Error Correction Code, but not really used to correct anything. I found sticking with "integrity protection" easier.

File: hw/ip/otbn/dv/model/otbn_rf_snooper_if.sv [hunk 10]
```diff
@@ -7,17 +7,22 @@
 
 `ifndef SYNTHESIS
 interface otbn_rf_snooper_if #(
-  parameter int Width = 32,    // Memory width in bits
-  parameter int Depth = 32     // Number of registers
+  parameter int Width = 39,      // Memory width in bits (including ECC bits)
+  parameter int ECCGranule = 32, // Number of data bits per seperate ECC, set to 0 if no ECC
```
[the remark this answers is not in the document]
Comment (GregAC) [reply hw/ip/otbn/dv/model/otbn_rf_snooper_if.sv:11]: I realise my "set to 0" comment was incorrect, you should actually just set `Width`, `DataWidth` and `ECCGranule` to the same setting (or set it to some divisor of `Width`) for things to work without ECC.

However as you say the granule is pretty baked into OTBN so I can probably move most of these as `localparam` within the module. I think I'll leave `DataWidth` and `Depth` with a new `IntegrityEnabled` as the module parameters.

File: hw/ip/otbn/rtl/otbn_rf_bignum.sv [hunk 11]
```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * 256b General Purpose Register File (GPRs) with ECC double error detection on 32-bit granule (312
```
[the remark this answers is not in the document]
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_rf_bignum.sv:6]: Indeed, had realised that after the initial PR!

**Review by GregAC — COMMENTED** [review]
Thanks for the review @imphil 

> Naming: I find "ECC" always confusing in the context here. Yes, we use an Error Correction Code, but not really used to correct anything.

This is true, I guess I just take "ECC" to mean Hamming/Hsaio codes and don't mentally unpack the full acronym.

I'll switch the language to talk about error detecting integrity codes (EDIC/EDC?). I believe we're actually using Hsaio codes at present, though there's scope to switch that around if desired so I think I'll avoid calling them Hamming codes directly.

**Comment by imphil** [comment]
I now also pushed the PR with the doc changes that I mentioned on Friday to https://github.com/lowRISC/opentitan/pull/6469. I'm not convinced I found the best possible naming there, maybe we can find good terminology over there and then apply it to the code?

File: hw/ip/otbn/dv/model/otbn_core_model.sv [hunk 12]
```diff
@@ -146,9 +146,25 @@ module otbn_core_model
...
     // syntax plus upwards name referencing to work with Verilator. Obviously, this won't work with
     // multiple OTBN instances, so it would be nice to get it right.
-    bind otbn_rf_base_ff otbn_rf_snooper_if #(.Width (32), .Depth (32)) u_snooper (.rf (rf_reg));
-    bind otbn_rf_bignum_ff otbn_rf_snooper_if #(.Width (256), .Depth (32)) u_snooper (.rf (rf));
-    bind otbn_rf_base otbn_stack_snooper_if #(.StackWidth (32), .StackDepth(8))
+    bind otbn_rf_base_ff otbn_rf_snooper_if #(
+      .Width (39),
+      .ECCGranule (32),
+      .DataWidth(32),
+      .Depth (32)
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply hw/ip/otbn/dv/model/otbn_core_model.sv:153]: Also, I'd consider removing the parameters and/or derive them from the bind site. Here, I think that `Width` can be `BaseECCWidth`. You can probably just hard-code `ECCGranule` and `DataWidth` as localparams in `otbn_rf_snooper_if`, since they are fundamental parts of the design. `Depth` is `NGpr` here.

Same comments for the binds below.

File: hw/ip/otbn/dv/tracer/rtl/otbn_trace_if.sv [hunk 13]
```diff
@@ -114,15 +114,29 @@ interface otbn_trace_if
 
   logic [WLEN-1:0] rf_bignum_wr_old_data;
 
-  if (RegFile == RegFileFF) begin : g_rf_ff_bignum_wr_old_data
-    assign rf_bignum_wr_old_data = gen_rf_bignum_ff.u_otbn_rf_bignum.rf[rf_bignum_wr_addr];
-  end else if (RegFile == RegFileFPGA) begin : g_rf_fpga_bignum_wr_old_data
-    assign rf_bignum_wr_old_data = gen_rf_bignum_fpga.u_otbn_rf_bignum.rf[rf_bignum_wr_addr];
+  logic [ExtWLEN-1:0] bignum_rf [NWdr];
+
+  for (genvar i = 0;i < NWdr; ++i) begin : g_probe_bignum_rf
```
Comment (rswarbrick) [inline hw/ip/otbn/dv/tracer/rtl/otbn_trace_if.sv:119]: Nit: Missing space after semicolon, and same on the next `for` loop.

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 14]
```diff
@@ -504,6 +526,15 @@ module otbn_controller
...
   assign mac_bignum_en_o = insn_executing & insn_dec_bignum_i.mac_en;
 
+  // Move / Conditional Select. Only select B register data when a selection instruction is being
+  // executed and the selection flag isn't set.
+
+  `ASSERT(SelFlagValid, insn_valid_i & insn_dec_bignum_i.sel_insn |->
+    insn_dec_bignum_i.alu_sel_flag inside {FlagC, FlagL, FlagM, FlagZ})
+
+  assign selection_result =
+    ~insn_dec_bignum_i.sel_insn | alu_bignum_selection_flag_i ? rf_bignum_rd_data_ecc_a_i :
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn_controller.sv:536]: Nit: Maybe this would be clearer with parens around the test? I had to think a bit about operator precedence here.

File: hw/ip/otbn/rtl/otbn_rf_bignum_ff.sv [hunk 15]
```diff
@@ -16,17 +19,17 @@ module otbn_rf_bignum_ff
...
+  output logic [ExtWLEN-1:0] rd_data_a_o,
 
-  input  logic [WdrAw-1:0] rd_addr_b_i,
-  output logic [WLEN-1:0]  rd_data_b_o
+  input  logic [WdrAw-1:0]   rd_addr_b_i,
+  output logic [ExtWLEN-1:0] rd_data_b_o
 );
-  logic [WLEN-1:0] rf [NWdr];
+  logic [ExtWLEN-1:0] rf [NWdr];
   logic [1:0]      we_onehot [NWdr];
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn_rf_bignum_ff.sv:33]: Nit: Indentation

**Comment by imphil** [comment]
Looks good, thanks Greg! (There is an open discussion at #6478  about the *inverted* Hsiao code mentioned in the comments, but let's sort that out once the discussion has concluded there.)

**Comment by imphil** [comment]
Can you rebase on top of master now that https://github.com/lowRISC/opentitan/pull/6476 is in? It will require changes to the prim_secded instantiations.
